### PR TITLE
Generalize format_diff_summary for C/C++/CUDA languages

### PR DESCRIPTION
**Pull request by odelliab**
## Summary

`format_diff_summary` (used to populate `metadata.changes` on programs) only recognized
Python `def`/`class` patterns. For non-Python code (e.g. CUDA kernels), it fell through
to a generic fallback that truncated the first line at 50 characters resulting in descriptions of the form "Change 1: Near #include <cuda.h>... (118→76 lines)"

This commit extends _extract_def_info to also recognize:
- CUDA: __global__, __device__, __host__ function signatures (including __launch_bounds__ qualifiers)
- C/C++: void/int/float/etc function signatures (with static/inline qualifiers)
- C++: struct/class definitions

Adds _extract_c_comment() for extracting comments from C-family code (block comments, line comments before or inside function bodies).

Generalizes _extract_first_comment() to handle // and /* */ in addition to Python # comments, and to find function bodies delimited by { in addition to Python's :.

Lastly, added a single constant (`_MAX_COMMENT_LINES = 5`) controlling comment truncation and a method 
(`_truncate_comment()) applying the limit at one place

After this change, CUDA diffs produce summaries like:

  "Change 1: Modified kernel fused_layer_norm_sm80: One block per row; blockDim=256, hidden=128.
Warps 0-3 active (tid 0-127), warps 4-7 idle (tid 128-255).
smem[0..3]=warp sums, [4..7]=warp sq, [8]=mean, [9]=inv_std"

Includes comprehensive test suite (31 tests) covering Python, CUDA, C/C++ patterns, and existing diff parsing functionality.

## Backward compatibility
- Python behavior is unchanged (same code paths, same output)
- The fallback path for unrecognized languages still works as before

### CODE REVIEW

File: skydiscover/utils/code_utils.py [hunk 1]
```diff
@@ -84,31 +86,73 @@ def parse_full_rewrite(llm_response: str, language: str = "python") -> Optional[
...
+        return ("class", name, _truncate_comment(comment))
+
+    # --- C/C++/CUDA ---
+    cuda_kernel_match = re.search(
+        r"^\s*(?:__global__|__device__|__host__(?:\s+__device__)?)\s+"
+        r"(?:__launch_bounds__\([^)]*\)\s*)?"
+        r"(?:\w+\s+)*?(\w+)\s*\(",
+        solution,
+        re.MULTILINE,
+    )
```
[the remark this answers is not in the document]
Comment (odelliab) [reply skydiscover/utils/code_utils.py:134]: CUDA regex — __launch_bounds__ now supported both before and after the return type

File: skydiscover/utils/code_utils.py [hunk 2]
```diff
@@ -84,31 +86,73 @@ def parse_full_rewrite(llm_response: str, language: str = "python") -> Optional[
...
+        comment = _extract_c_comment(solution, cuda_kernel_match.start())
+        return (kind, name, _truncate_comment(comment))
+
+    c_func_match = re.search(
+        r"^\s*(?:static\s+|inline\s+|__inline__\s+)*"
+        r"(?:void|int|float|double|half|bool|char|unsigned|long|size_t|auto|\w+_t)\s+"
+        r"(?:\*\s*)?(\w+)\s*\(",
+        solution,
+        re.MULTILINE,
+    )
```
[the remark this answers is not in the document]
Comment (odelliab) [reply skydiscover/utils/code_utils.py:149]: C function regex — added const qualifier and pointer stars adjacent to type (char**)

File: tests/utils/test_code_utils.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+void fused_layer_norm(const half* x, half* y, int hidden, float eps) {
+    // One block per row; blockDim=256
+    const int row = blockIdx.x;
+}"""
+        result = _extract_def_info(code)
+        assert result is not None
+        kind, name, docstring = result
+        assert kind == "kernel"
+        assert name == "fused_layer_norm"
+        assert "One block per row" in docstring
```
[the remark this answers is not in the document]
Comment (odelliab) [reply tests/utils/test_code_utils.py]: added new  test: _launch_bounds_after_return_type

File: tests/utils/test_code_utils.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+        code = """static inline float fast_rsqrt(float x) {
+    // Quake III fast inverse square root
+    return 1.0f / sqrtf(x);
+}"""
+        result = _extract_def_info(code)
+        assert result is not None
+        kind, name, docstring = result
+        assert kind == "function"
+        assert name == "fast_rsqrt"
+        assert docstring == "Quake III fast inverse square root"
```
[the remark this answers is not in the document]
Comment (odelliab) [reply tests/utils/test_code_utils.py:181]: added new test: test_const_pointer_function

File: tests/utils/test_code_utils.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        code = """void func() {
+    /*
+     * Step 1: load data
+     * Step 2: compute
+     */
+    int x = 0;
+}"""
+        result = _extract_first_comment(code, 0)
+        assert result is not None
+        assert "Step 1: load data" in result
```
[the remark this answers is not in the document]
Comment (odelliab) [reply tests/utils/test_code_utils.py:254]: is_c_style variable is not needed as the implementation auto-detects comment style

**Comment by shubham3-ucb** [comment]
@odelliab nice generalization, and the test coverage is great (33 passing locally). Since this only feeds the `metadata.changes` summary string the blast radius is small, but two things before the merge is approved:

1. Lint is red. black wants to reformat `skydiscover/utils/code_utils.py`. A quick `black skydiscover/` and a push should turn CI green (tests already pass).

2. Minor Python edge case. The new `*` branch in `_extract_first_comment` (meant for C block-comment continuation lines) also catches Python lines that start with `*`. For a docstring-less function whose first body line is an unpack like `*rest, last = ...`, the summary now grabs that as a comment:

   ```
   def k():
       *rest, last = [1, 2, 3]
       return last
   # main:    ('function', 'k', None)
   # this PR: ('function', 'k', 'rest, last = [1,2,3]')
   ```

   It is only the human-readable summary so nothing functional breaks, but it is a behavior change for Python. Could the `*` continuation branch be scoped to the C/CUDA paths (or to when we are already inside a `/* */` block) so Python is untouched?

Otherwise Python output matches main on everything I checked. This should be good to merge once lint is green and the `*` case is handled or confirmed fine.

**Comment by odelliab** [comment]
Thank you for the comments.
1. Fixed the python * edge-case and added a test to make sure a python line code starting with * will not be recognized as a comment. 
2. hopefully, fixed also the linting error

**Review by lynnliu030 — APPROVED** [review]
LGTM!
